### PR TITLE
KMAC buffer overflow fix

### LINKED ISSUE #14795: kmac_set_ctx_params() check is wrong.
If the customization string is 127 bytes it passes the check. Then `kmac_init()` calls `bytepad()`, with in1_len=6, in2_len=130, and w=136. `bytepad()` tries to pad the buffer with 134 zero bytes.
``` C
/* zero pad the end of the buffer */
memset(p, 0, sz - len);
```
The buffer can only hold 168 bytes and there are already 138 bytes in the buffer This is what the check in `kmac_set_ctx_params()` was trying to protect against.

### PR DESCRIPTION
**Pull request by paulidale**
Fixes a buffer overflow (two byte on the stack).
Also increases the length of the customisation string.

Fixes #14795

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: providers/implementations/macs/kmac_prov.c [hunk 1]
```diff
@@ -296,10 +307,12 @@ static int kmac_init(void *vmacctx, const unsigned char *key,
         (void)kmac_set_ctx_params(kctx, cparams);
     }
 
-    return bytepad(out, &out_len, kmac_string, sizeof(kmac_string),
-                   kctx->custom, kctx->custom_len, block_len)
-           && EVP_DigestUpdate(ctx, out, out_len)
-           && EVP_DigestUpdate(ctx, kctx->key, kctx->key_len);
+    res = bytepad(out, &out_len, kmac_string, sizeof(kmac_string),
```
Comment (slontis) [inline providers/implementations/macs/kmac_prov.c:310]: probably wouldnt hurt to assert that out_max >= out_len

File: providers/implementations/macs/kmac_prov.c [hunk 2]
```diff
@@ -286,6 +286,17 @@ static int kmac_init(void *vmacctx, const unsigned char *key,
     block_len = EVP_MD_block_size(ossl_prov_digest_md(&kctx->digest));
     if (block_len < 0)
         return 0;
+    /*
+     * The temporary buffer for the customisation and header needs to contain:
+     *      len(encoded_int(block_len)) + len(encoded_string(custom))
+     *      + len(kmac_encoded_string)
+     * bytes rounded up to the next block_size multiple.
+     */
+    out_max = ((2 + kctx->custom_len + sizeof(kmac_string) + block_len - 1)
```
Comment (slontis) [inline providers/implementations/macs/kmac_prov.c:295]: Normally I would make the bytepad function figure out this number if out is NULL..

File: test/recipes/30-test_evp_data/evpmac_common.txt [hunk 3]
```diff
@@ -400,3 +400,12 @@ Custom = "My Tagged Application"
 Output = D5BE731C954ED7732846BB59DBE3A8E30F83E77A4BFF4459F2F1C2B4ECEBB8CE67BA01C62E8AB8578D2D499BD1BB276768781190020A306A97DE281DCC30305D
 Ctrl = size:64
 Ctrl = xof:1
+
+Title = KMAC long customisation string
```
Comment (slontis) [inline test/recipes/30-test_evp_data/evpmac_common.txt:404]: Do you need to quote the source of the test vector?

File: providers/implementations/macs/kmac_prov.c [hunk 4]
```diff
@@ -286,6 +286,17 @@ static int kmac_init(void *vmacctx, const unsigned char *key,
...
+               / block_len) * block_len;
+    out = OPENSSL_malloc(out_max);
+    if (out == NULL)
+        return 0;
```
Comment (FdaSilvaYY) [inline providers/implementations/macs/kmac_prov.c:299]: Deserves a call to 
```
                    ERR_raise(ERR_LIB_..., ERR_R_MALLOC_FAILURE);
```

File: test/recipes/30-test_evp_data/evpmac_common.txt (hunk 3 above)
Comment (paulidale) [reply test/recipes/30-test_evp_data/evpmac_common.txt:404]: No, I created it and I'd like it replaced by a vector from NIST.

File: providers/implementations/macs/kmac_prov.c (hunk 4 above)
Comment (paulidale) [reply providers/implementations/macs/kmac_prov.c:299]: Yep, added.

File: providers/implementations/macs/kmac_prov.c (hunk 2 above)
Comment (paulidale) [reply providers/implementations/macs/kmac_prov.c:295]: Can't say I like this this approach, but it is used elsewhere....

Comment (paulidale) [reply providers/implementations/macs/kmac_prov.c:295]: And done.

File: providers/implementations/macs/kmac_prov.c (hunk 1 above)
Comment (paulidale) [reply providers/implementations/macs/kmac_prov.c:310]: No need with the extra call to bytepad() to figure out the size.

**Comment by paulidale** [comment]
I've also run through and converted the sizes to size_t.

File: test/recipes/30-test_evp_data/evpmac_common.txt (hunk 3 above)
Comment (paulidale) [reply test/recipes/30-test_evp_data/evpmac_common.txt:404]: I've added a vector from the ACVP and noted the source in the test data.

**Comment by paulidale** [comment]
Should be all happy for a re-review now.

**Review by slontis — COMMENTED** [review]
Looks good to me.. A few negative tests might be good.
e.g Custom string  = 256..

**Comment by paulidale** [comment]
Added an overly long customisation string test case (257 characters).

**Review by slontis — APPROVED** [review]
Approved if test pass..

**Comment by slontis** [comment]
Pauli can you try to remember to add the #fixes to the description at the top please...

**Comment by paulidale** [comment]
Merged to master, I'll try to remember to include the fixes.

**Pull request closed without merging**